### PR TITLE
DRIVERS-1807 Keep existing retry behavior unless CSOT is enabled

### PR DESCRIPTION
**Pull request by ShaneHarvey**
Please complete the following before merging:

- [x] Update changelog.
- [x] Make sure there are generated JSON files from the YAML test files.
- [x] Test changes in at least one language driver.
- [x] Test these changes against all server versions and topologies (including standalone, replica set, sharded clusters, and serverless).

### CODE REVIEW

File: source/client-side-operations-timeout/client-side-operations-timeout.rst [hunk 1]
```diff
@@ -47,8 +47,10 @@ timeoutMS
 ~~~~~~~~~
 
 This 64-bit integer option specifies the per-operation timeout value in
-milliseconds. The default value is unset. Both unset and an explicit value of
-0 mean infinite, though some client-side timeouts like
+milliseconds. The default value is unset which means this feature is not
+enabled, ie the existing timeout behavior is unchanged (including
```
Comment (benjirewis) [inline source/client-side-operations-timeout/client-side-operations-timeout.rst:51]:
```suggestion
enabled, i.e. the existing timeout behavior is unchanged (including
```

File: source/retryable-writes/retryable-writes.rst [hunk 2]
```diff
@@ -381,36 +382,35 @@ writes, drivers MUST add a transaction ID to the command. Drivers MUST only
...
-If the retry attempt also fails, drivers MUST update their topology according to
+If a retry attempt also fails, drivers MUST update their topology according to
 the SDAM spec (see: `Error Handling`_). If an error would not allow the caller
 to infer that an attempt was made (e.g. connection pool exception originating
-from the driver), the original error should be raised. If the retry failed due
-to another retryable error or some other error originating from the server, that
-error should be raised instead as the caller can infer that an attempt was made
-and the second error is likely more relevant (with respect to the current
-topology state).
+from the driver) or the error is labeled "NoWritesPerformed", the error from
```
Comment (benjirewis) [inline source/retryable-writes/retryable-writes.rst:408]: Is there any other case besides `NoWritesPerformed` where we return the previous error?

File: source/retryable-writes/retryable-writes.rst [hunk 3]
```diff
@@ -443,61 +443,65 @@ Consider the following pseudo-code:
...
-    try {
-      return executeCommand(server, retryableCommand);
-    } catch (DriverException ignoredError) {
-      throw originalError;
-    } catch (Exception secondError) {
-      handleError(secondError);
-      throw secondError;
+      /* CSOT is enabled and the operation has timed out. */
+      if (timeoutMS != null && is_expired(timeoutMS) {
+        throw previousError;
```
Comment (benjirewis) [inline source/retryable-writes/retryable-writes.rst:503]: Do we have a spec test for this block?

**Review by benjirewis — COMMENTED** [review]
Might want to add @prestonvasquez since some of this has to do with DRIVERS-2327

File: source/retryable-writes/retryable-writes.rst (hunk 2 above)
Comment (ShaneHarvey) [reply source/retryable-writes/retryable-writes.rst:408]: Yes, when a retry selects a server that does not support retryable writes or the retry fails with a server selection error.

File: source/retryable-writes/retryable-writes.rst (hunk 3 above)
Comment (ShaneHarvey) [reply source/retryable-writes/retryable-writes.rst:503]: Yes we do, it's part of the CSOT. The "timeoutMS applies to whole operation, not individual attempts" tests: https://github.com/mongodb/specifications/blob/b3e06831a09ae72d94da79f428d8941ec7221f7f/source/client-side-operations-timeout/etc/templates/retryability-timeoutMS.yml.template#L42-L47

**Review by benjirewis — APPROVED** [review]
Thanks!

**Review by prestonvasquez — APPROVED** [review]
LGTM!

**Review by isabelatkinson — APPROVED** [review]

**Comment by ShaneHarvey** [comment]
I've updated this PR to include similar changes to the retryable reads spec. I've also adjusted some of the CSOT spec tests to be less flakey (mainly by increasing timeouts and using w:1 instead of w:majority). Ready for another look.

**Review by isabelatkinson — APPROVED** [review]
lgtm!

**Review by benjirewis — APPROVED** [review]
CSOT test changes LGTM and pass in the Go driver.

**Pull request merged**